src/dfx/src/commands/deploy.rs

```diff
@@ -183,13 +212,27 @@ fn display_urls(env: &dyn Environment) -> DfxResult {
                             "https://{}.raw.ic0.app/?id={}",
                             MAINNET_CANDID_INTERFACE_PRINCIPAL, canister_id
                         );
-                        candid_urls.insert(canister_name, Url::parse(&url)?);
+                        candid_urls.insert(
+                            canister_name,
+                            Url::parse(&url).with_context(|| {
+                                format!(
+                                    "Failed to parse candid url {} for canister {}.",
+                                    &url, canister_name
+                                )
+                            })?,
+                        );
                     } else if let Some(ui_canister_id) = ui_canister_id {
-                        let mut url = Url::parse(&network.providers[0])?;
+                        let mut url = Url::parse(&network.providers[0]).with_context(|| {
```

Comment:
Yeah, I think it makes sense to extract such a block to a method, and I just did so in the latest commit. But I think this is a general problem we have: We accept very deeply nested things in our code and are reluctant/too lazy to just factor out some functions, even though it would increase readability by quite a lot.